---
layout: page
title: 2018 Project ideas list
permalink: /project-ideas/2018/
---

Quick links:
 - [Cross-InterMine search tool](#cross-intermine-search-tool)
 - [Data browser tool](#data-browser-tool)
 - [Python client expansion](#python-client-expansion)
 - [InterMineR package expansion](#interminer-package-expansion)
 - [InterMine Neo4j database and Bluegenes integration](#intermine-neo4j-database-and-bluegenes-integration)
 - [InterMine ElasticSearch/Solr deep bio data search](#intermine-elasticsearchsolr-deep-bio-data-search)
 - [Buzzbang Google-like bio data search](#buzzbang-google-like-bio-data-search)
 - [Automated test suite for ClojureScript application](#automated-test-suite-for-clojurescript-application)
 - [Query builder structured language to natural language translator](#query-builder-structured-language-to-natural-language-translator)

# Projects

## Cross-InterMine search tool
There are lots of InterMines based in different locations around the world. It would be nice to use [InterMine WebServices](http://iodocs.apps.intermine.org/) to type a search term once, and search many (or all) InterMines at the same time, then display all results sets on the same page. Would use the InterMine registry ([ui](http://registry.intermine.org/), [api docs](http://registry.intermine.org/api-docs/), [api](http://registry.intermine.org/service/instances)) to pull an up-to-date list of InterMines.  

#### Project results:
A client-driven cross-InterMine search page.

#### Skills required:
HTML, CSS, JavaScript, (or some language that compiles to JavaScript). No biological knowledge required.

#### Skill level:
Medium.

#### Mentors:
- Yo Yehudi - yo@intermine.org
- Nadia Yudina - joystate@me.com

## Data browser tool
InterMine needs a faceted search tool to display data. E.g something like https://dcc.icgc.org/search. A possible (but non-binding) mockup of how this might look for InterMine is available here: https://www.lucidchart.com/documents/view/3cd706c4-2c71-49fa-b8e7-c724432c26e4/0

#### Project results:
A tool to browse InterMine data
Skills required: HTML, CSS, JavaScript, (or some language that compiles to JavaScript). No biological knowledge required.

#### Skill level:
Medium.

#### Mentors:
- Yo Yehudi - yo@intermine.org
- Nadia Yudina joystate@me.com

## Python client expansion
InterMine has a client in Python with enables the querying of InterMine databases via Python scripts. However this client does not include all functionality available in InterMine so we’d like to extend the library to include features such as the region search and enrichment. [See here for project details](https://docs.google.com/document/d/1KI3f2ArHAbssVV1g6L7-Oxk7CatK2r7iH0inoUrbgfQ/edit?usp=sharing)

#### Project results:
Improved InterMine Python client

#### Skills required:
Python, No biological knowledge required.

#### Skill level:
Medium.

#### Mentors:
- Julie Sullivan - julie@intermine.org
- Rachel Lyne - rachel@intermine.org

## InterMineR package expansion
InterMineR package provides an interface of the R programming environment with InterMine databases. The package facilitates the performance of complex queries and enrichment analysis with R, and has been accepted in Bioconductor. However, there are still many features that can be added to improve its functionality and usability. Therefore we’d like to create a Shiny-based UI, to facilitate the usage of InterMineR, and build-in functions for visualization and statistical analysis based on the ggplot2 and igraph R packages.

#### Project results:
Improved InterMineR package

#### Skills required:
- R programing language

#### Skill level:
Medium.
#### Mentors:
- Rachel Lyne - rachel@intermine.org
- Julie Sullivan - julie@intermine.org

## InterMine Neo4j database and BlueGenes (new UI) integration
InterMine is an open source integration system that uses PostgreSQL to store complex biological data. The Neo4j project (https://github.com/intermine/neo4j) already replaces the current relational database with Neo4j graph database; it includes the data loaders for InterMine Neo4j database and the RESTful API from which you can query InterMine Neo4j database using PathQuery. This project might be extended by:  

- implementing a full integration with InterMine's new user interface
- improving the data loaders
- providing some benchmark results

#### Project results:
Fully-functional InterMine Neo4j database running from InterMine's new user interface

####  Skills required
- Good understanding of Java
- Good understanding of RESTful API
- Basic understanding of graph databases
- Git or other version control
- No biology skills needed

#### Mentors:
- Sam Hokin  - shokin@ncgr.org
- Daniela Butano - daniela@intermine.org

## InterMine ElasticSearch/Solr deep bio data search
At the moment, we use an ancient version of Lucene to index the contents of a mine in a primitive way (i.e. we put almost every field into the index and do some simple weighting). This is stuffed into InterMine’s own Postgres database and rehydrated at runtime. We want to use a modern search system so your mission (should you choose to accept it) would be to:

- Assess ElasticSearch (ES) and Solr (especially BioSolr) as to which would be better to include with InterMine. We (the mentors) will make a joint assessment with you as to which to use.
- Re-code and re-architect InterMine to use ES/Solr instead of embedded Lucene. This means managing a separate process whilst keeping the system simple for a non-developer to set up and use.
- Explore how ES/Solr can provide more sophisticated search for InterMine (e.g ontology enhanced search via BioSolr).

#### Project results:
InterMine uses a modern search system which improved searchability.

#### Skill level:
Hard

#### Skills required
- Java
- Docker or a strong interest to learn
- Solr/ES or a strong interest to learn
- Knowledge of ontologies as used in academic software engineering, esp. biological ones, or a strong interest to learn
#### Mentors
- Daniela Butano - daniela@intermine.org
- Justin Clark-Casey - justincc@intermine.org

## Buzzbang Google-like bio data search
Life sciences produce a vast amount of data with a very wide range, encompassing everything from gene and protein sequences to biobank samples and bioinformatics training courses. This data is also distributed amongst a multitude of websites, making it hard to search. A new initiative called Bioschemas provides a way to embed structured markup in webpages, which we can then crawl to provide a Google-like search over the biological web.  
### Tasks:
- Improve the mechanisms for embedding Bioschemas markup in InterMine webpages
- Help us improve the Buzzbang Bioschemas crawler and indexer
- Help us improve the Buzzbang search frontend.

#### Project results:
InterMine has a generic way to embed Bioschemas markup and improved mechanisms for crawling, indexing and searching such markup.

#### Skill level:
Hard

#### Skills required
- Java
- Python
- Solr
- Flask

#### Mentors
- Justin Clark-Casey - justincc@intermine.org
- Needs another possible mentor!

## Automated test suite for ClojureScript application
BlueGenes ([GitHub](https://github.com/intermine/bluegenes), [Demo](http://bluegenes.apps.intermine.org/)) is a [ClojureScript](https://clojurescript.org/) browser-based UI for [InterMine](http://intermine.org/) that is lacking in good tests. Currently, before merging a pull request, reviewers need to manually validate certain pages work correctly using a [checklist](https://github.com/intermine/bluegenes/blob/dev/PULL_REQUEST_TEMPLATE.md). Ideally many of these checks could be automated into a test suite to save reviewers time and making it harder to miss errors!

#### Project results:
A headless (e.g. phantomjs or similar) UI test suite using a mock server for http requests. Stretch goal: hooked up to [Travis](https://travis-ci.org/) or similar CI.

#### Skills required:
- Clojurescript / Clojure
- Advantageous: ability to examine software and identify+implement useful test cases.

#### Skill level:
Medium.

#### Mentors:
- Yo Yehudi yo@intermine.org
- Josh Heimbach josh@intermine.org

## Query builder structured language to natural language translator
InterMine uses a structured GUI-based query builder ([demo](http://www.flymine.org/flymine/customQuery.do)) to allow users to construct queries into complex biological data. Whilst InterMine experts are comfortable building queries with this interface, it's not entirely intuitive and can seem intimidating to newer users, despite [extensive help  documentation](http://flymine.readthedocs.io/en/latest/query-builder/Documentationquerybuilder.html). Given that all queries _can_ be expressed in natural language, we would like to build a library that can translate from an InterMine query to a natural language statement.

#### Project results:
Given any InterMine [PathQuery](http://intermine.readthedocs.io/en/latest/api/pathquery/), return a natural language sentence describing the query.

#### Skills required:
- ClojureScript / Clojure OR JavaScript

#### Skill level:
Medium.

#### Mentors:
- Yo Yehudi yo@intermine.org
- Josh Heimbach josh@intermine.org
